### PR TITLE
[EH] Make getCppExceptionThrownValue return thrown value

### PR DESCRIPTION
**Pull request by aheejin**
In Emscripten EH, the thrown value by the JS library is the original
user-thrown value itself.

But in Wasm EH, what we throw is an [`WebAssembly.Exception`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/Exception) object. You
can use getArg method to get the value thrown. But even this value is
not the actual user-thrown value; in libc++abi, which Wasm EH uses, what
we actually throw is a pointer to something called unwind header,
represented by `_Unwind_Exception`:
https://github.com/emscripten-core/emscripten/blob/1f6b136d75c9a6a89d07ed069b9cede80b39cebd/system/lib/libcxxabi/src/cxa_exception.cpp#L284

The current `getCppExceptionThrownValue` in our JS exception library
returns the `_Unwind_Exception` pointer, which is the value actually
thrown by libc++abi, and not the original user-thrown value. This is
confusing in the first place, and I got requests from users who want to
access the original user-thrown value.

Because of `WebAssembly.Exception` object, we still have some
discrepancy between Emscripten EH and Wasm EH, but I think this is more
intuitive, and it actually simplifies libc++abi code as well.

So the current state:
- Emscripten EH: Throws original user-thrown value
- Wasm EH: Throws `WebAssembly.Exception` object, which contains the
           pointer to the unwind header

After this PR:
- Emscripten EH: Throws original user-thrown value
-  Wasm EH: Throws `WebAssembly.Exception` object, which contains the
          original user-thrown value

Addresses #16033.

### CODE REVIEW

**Comment by aheejin** [comment]
cc @hoodmane @Krelborn @donalffons

File: src/library_exceptions.js [hunk 1]
```diff
@@ -398,29 +398,32 @@ var LibraryExceptions = {
...
+  $decrementExceptionRefcount__deps: ['__cxa_decrement_exception_refcount', '$getCppExceptionThrownValue'],
   $decrementExceptionRefcount: function(obj) {
     var ptr = getCppExceptionThrownValue(obj);
-    ___decrement_wasm_exception_refcount(ptr);
+    ___cxa_decrement_exception_refcount(ptr);
   },
 
   $getExceptionMessage__deps: ['__get_exception_message', 'free', '$getCppExceptionThrownValue'],
-  $getExceptionMessage: function(ptr) {
+  $getExceptionMessage: function(obj) {
```
Comment (aheejin) [inline src/library_exceptions.js:422]: Drive-by fix: the variable names (`obj` and `ptr`) were swapped in the previous code.

File: system/lib/libcxxabi/src/cxa_exception_emscripten.cpp [hunk 2]
```diff
@@ -33,27 +33,26 @@ thrown_object_from_cxa_exception(__cxa_exception* exception_header) {
 //  Get the exception object from the unwind pointer.
 //  Relies on the structure layout, where the unwind pointer is right in
 //  front of the user's exception object
-static inline __cxa_exception* cxa_exception_from_exception_unwind_exception(
+static inline __cxa_exception* cxa_exception_from_unwind_exception(
```
Comment (aheejin) [inline system/lib/libcxxabi/src/cxa_exception_emscripten.cpp:36]: This was named `cxa_exception_from_exception_unwind_exception` because it was copied from cxa_exception.cpp. I'm not sure why we need to repeat `exception`, so I shortened the name. The same with the function below.

**Comment by aheejin** [comment]
Gentle ping 😀

File: system/lib/libcxxabi/src/cxa_exception_emscripten.cpp (hunk 2 above)
Comment (sbc100) [reply system/lib/libcxxabi/src/cxa_exception_emscripten.cpp:36]: If this is more correct we should maybe propose this as an upstream change?    I do think there is some value in matching these names with the upstream ones... but I don't feel strongly about it.

File: src/library_exceptions.js [hunk 3]
```diff
@@ -398,29 +398,32 @@ var LibraryExceptions = {
     return Module['asm']['__cpp_exception'];
   },
 
-  $getCppExceptionThrownValue__deps: ['$getCppExceptionTag'],
+  $getCppExceptionThrownValue__deps: ['$getCppExceptionTag', '__thrown_object_from_unwind_exception'],
   $getCppExceptionThrownValue: function(ex) {
```
Comment (sbc100) [inline src/library_exceptions.js:402]: Should we call this function `getCppExceptionThrownObject`?  Or is thrown value a thing that is different?

**Review by sbc100 — APPROVED** [review]
lgtm % comments

File: src/library_exceptions.js (hunk 3 above)
Comment (aheejin) [reply src/library_exceptions.js:402]: Yeah these are confusing. In library_exceptions.js, I tried to use 'object' for `WebAssembly.Exception` object thrown, and 'value' for the user-thrown value. So this takes an object and returns the user-thrown value. This got more confusing by the param name for this function is `ex` while other functions' param name is `obj`, when they denote the same thing.

At the same time libc++abi uses 'object' to mean the user-thrown value, which adds to the confusion.. I guess we should make these terms clearer.

How about this?
- object: the thrown value. So we rename this function to `getCppExceptionThrownObject` as you suggested.
- value: let's not use this term at all
- ex/exception: `WebAssembly.Exception`. So we rename current `obj` params in other functions, which meant `WebAssembly.Exceptions`, to `ex`.

If you think this sound fine, I'll make this change as a followup PR.

Comment (sbc100) [reply src/library_exceptions.js:402]: Maybe "JSObject" or "JSException"?

Comment (aheejin) [reply src/library_exceptions.js:402]: What is the thing you want to represent with "JSObject" or "JSException"? `WebAssembly.Exception` or the user-thrown value(or object)? I guess the latter? In that case it's confusing in another way because this is a Wasm exception and that sounds like this is a JS exception thrown from JS, a foreign exception..

Comment (sbc100) [reply src/library_exceptions.js:402]: `WebAssembly.Exception` is what I propose to refer to as JS object.   The thrown object lives in wasm memory, as does the exception object, right?  They are both allocated with malloc.

Comment (sbc100) [reply src/library_exceptions.js:402]: This is how I've been thinking about it:   JS Object (`WebAssembly.Exception`) -> refers to a C++ `Exception` class object allocated by libc++ -> refers to a user object that that user threw.  

JSExceptionObject -> C++ExceptionObject -> UserObject/ThrownValue

Comment (aheejin) [reply src/library_exceptions.js:402]: Sorry I'm not sure if I understand. `WebAssembly.Exception` is not what the user threw, and this is the source of all these term differences. 

> JSExceptionObject -> C++ExceptionObject -> UserObject/ThrownValue

But this sounds like these three are the same? Is that what you mean?

Comment (sbc100) [reply src/library_exceptions.js:402]: The user throws a C++ object of their choosing (the thrown object), libc++ then wraps then into an Exception object (still native pointer to memory), and then this can be caught on the outside as a JavaScript exception object.

Is that about right?

They do all kind of all refer to same thing, but they are distinct right?     And in particular that is no way to pass the JSObject into native code (since its not a number like the other two).

Comment (aheejin) [reply src/library_exceptions.js:402]: The user throws a C++ object (native pointer), and what libc++abi throws is a pointer to `_Unwind_Exception`, which still a native pointer but an address preceding the user-thrown project. In this layout, the start address of the third block (thrown object) is what an user originally throws, and the start address of `_Unwind_Exception` is what libc++abi ends up throwing.
https://github.com/emscripten-core/emscripten/blob/24f765dc3545e89f232c9f8503f6426055271628/system/lib/libcxxabi/src/cxa_exception.cpp#L26-L28

`WebAssembly.Exception` is something different. This has nothing to do with libc++abi. libc++abi calls libunwind to do the actual low level throwing stuff from the VM: https://github.com/emscripten-core/emscripten/blob/24f765dc3545e89f232c9f8503f6426055271628/system/lib/libcxxabi/src/cxa_exception.cpp#L284

And our libunwind port calls `__builtin_wasm_throw`, which is lowered down to a wasm `throw` instruction: https://github.com/emscripten-core/emscripten/blob/24f765dc3545e89f232c9f8503f6426055271628/system/lib/libunwind/src/Unwind-wasm.c#L50-L56

And within this Wasm native `throw` instruction, it wraps the value thrown by libc++abi (a pointer to `_Unwind_Exception`) into a `WebAssembly.Exception` object, and throws that object.

Nevermind `_Unwind_Exception` thing here, which I don't want to expose. What we are confused about is the use of the term 'object'. Does it denote the user-thrown native pointer, or `WebAssembly.Exception` object?

I still think `JSException` is confusing because this sounds like a foreign (non-wasm) exception..

Comment (sbc100) [reply src/library_exceptions.js:402]: I see, so `WebAssembly.Exception`, despite being a JavaScript object, starts life on the native/wasm side, so it corresponds to a wasm object?   On the wasm side its a tag/i32 pair or something like that?

However in this code it really is a JS heap object, right?    This function take a JS heap object and returns a pointer/number, right?

Comment (sbc100) [reply src/library_exceptions.js:402]: For example a longer name for this function might be something like `getCppThrownObjectFromJSCaughtObject`?

Comment (aheejin) [reply src/library_exceptions.js:402]: Not sure `JSCaughtObject` is clearer to someone who doesn't know this context. How about just `getCppThrownObjectFromWebassemblyException`, with comments clarifying what this means? This name is long but I guess at least not misleading..

Comment (sbc100) [reply src/library_exceptions.js:402]: Sure, sounds reasonable.   I think if was a user I would be looking for something do with the thing that I caught.. so I guess `JSCaughtObject` and `WebassemblyException` both describe that thing.

Comment (aheejin) [reply src/library_exceptions.js:402]: I'll explain in the comment that this is `WebAssembly.Exception` object that Wasm EH JS API provides.

Comment (aheejin) [reply src/library_exceptions.js:402]: Anyway, will do this in a followup. Thanks for the discussion!

**Comment by aheejin** [comment]
Will land this for now, but I'm planning on changing some methods' names in library_exception.js, so please don't consider the current APIs as settled just yet!
